prml/doughnut/src/impls.rs

```diff
@@ -0,0 +1,319 @@
+		let holder = sr25519::Pair::from_string("//Bob", None).unwrap();
+		let doughnut = DoughnutV0 {
+			issuer: issuer.public().into(),
+			holder: holder.public().into(),
+			expiry: 3000,
+			not_before: 0,
+			payload_version: 0,
+			signature: [1u8; 64].into(),
+			signature_version: 0,
+			domains: vec![("test".to_string(), vec![0u8])],
+		};
+		let plug_doughnut = PlugDoughnut::<_, Runtime>::new(doughnut);
+		assert!(
+			<PlugDoughnut<_, _> as DoughnutApi>::validate(&plug_doughnut, holder.public(), 100).is_ok()
+		);
+	}
+
+	#[test]
+	fn plug_doughnut_does_not_validate() {
+		let issuer = sr25519::Pair::from_string("//Alice", None).unwrap();
+		let holder = sr25519::Pair::from_string("//Bob", None).unwrap();
+		let signer = sr25519::Pair::from_string("//Charlie", None).unwrap();
+		let doughnut = DoughnutV0 {
+			issuer: issuer.public().into(),
+			holder: holder.public().into(),
+			expiry: 3000,
+			not_before: 1000,
+			payload_version: 0,
+			signature: [1u8; 64].into(),
+			signature_version: 0,
+			domains: vec![("test".to_string(), vec![0u8])],
+		};
+		let plug_doughnut = PlugDoughnut::<_, Runtime>::new(doughnut);
+		// premature
+		assert!(
+			<PlugDoughnut<_, _> as DoughnutApi>::validate(&plug_doughnut, holder.public(), 999).is_err()
+		);
+		// expired
+		assert!(
+			<PlugDoughnut<_, _> as DoughnutApi>::validate(&plug_doughnut, holder.public(), 3001).is_err()
+		);
+		// signer is not holder
+		assert!(
+			<PlugDoughnut<_, _> as DoughnutApi>::validate(&plug_doughnut, signer.public(), 100).is_err()
+		);
+	}
+
+	#[test]
+	fn plug_doughnut_verifies_sr25519_signature() {
+		let issuer = sr25519::Pair::from_string("//Alice", None).unwrap();
+		let holder = sr25519::Pair::from_string("//Bob", None).unwrap();
+		let mut doughnut = DoughnutV0 {
+			issuer: issuer.public().into(),
+			holder: holder.public().into(),
+			expiry: 0,
+			not_before: 0,
+			payload_version: 0,
+			signature: [1u8; 64].into(),
+			signature_version: 0,
+			domains: vec![("test".to_string(), vec![0u8])],
+		};
+		doughnut.signature = issuer.sign(&doughnut.payload()).into();
+		let plug_doughnut = PlugDoughnut::<_, Runtime>::new(doughnut);
+		assert!(plug_doughnut.verify());
+	}
+
+	#[test]
+	fn plug_doughnut_does_not_verify_sr25519_signature() {
+		let issuer = sr25519::Pair::from_string("//Alice", None).unwrap();
+		let holder = sr25519::Pair::from_string("//Bob", None).unwrap();
+		let mut doughnut = DoughnutV0 {
+			issuer: issuer.public().into(),
+			holder: holder.public().into(),
+			expiry: 0,
+			not_before: 0,
+			payload_version: 0,
+			signature: [1u8; 64].into(),
+			signature_version: 0,
+			domains: vec![("test".to_string(), vec![0u8])],
+		};
+		doughnut.signature = holder.sign(&doughnut.payload()).into();
+		let plug_doughnut = PlugDoughnut::<_, Runtime>::new(doughnut);
+		assert_eq!(plug_doughnut.verify(), false);
+	}
+
+	#[test]
+	fn plug_doughnut_verifies_ed25519_signature() {
```

Comment:
this test fails on my machine

Reply:
I'm trying to fix these at the moment. It's a weird build error, it works if you run it like `cargo test -p prml-doughnut`